File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-3.cpp
class Solution {
public:
    int findMin(vector<int> &nums) {
        
        int l = 0;
        int r = nums.size() - 1;
        int ans = nums[0];
        while(l<=r){
            if(nums[l] <= nums[r]){
                ans = min(nums[l],ans);
            }
            int mid = l + (r-l) / 2;
            ans = min(nums[mid],ans);
            if(nums[l] <= nums[mid]){
                l = mid+1;
            }
            else{
                r = mid - 1;
            }

        }
        return ans;
    }
};

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-3.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1

        while l <= r:
            m = (l+r) // 2
            if nums[m] == target:
                return m
            elif nums[m] >= nums[r]:
                if nums[l] <= target and target < nums[m]:
                    r = m - 1
                else:
                    l = m + 1
            else:
                if nums[r] >= target and target > nums[m]:
                    l = m + 1
                else:
                    r = m - 1
        
        return -1
